src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -149,45 +193,116 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
+            {
+                if (protocols == SslProtocols.None)
+                {
+                    // we are using default settings but cipher suites policy says that TLS 1.3
+                    // is not compatible with our settings (i.e. we requested no encryption or disabled
+                    // all TLS 1.3 cipher suites)
+                    protocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12;
+                }
+                else
+                {
+                    // user explicitly asks for TLS 1.3 but their policy is not compatible with TLS 1.3
+                    throw new SslException(
+                        SR.Format(SR.net_ssl_encryptionpolicy_notsupported, sslAuthenticationOptions.EncryptionPolicy));
                 }
+            }
 
+            if (CipherSuitesPolicyPal.ShouldOptOutOfLowerThanTls13(sslAuthenticationOptions.CipherSuitesPolicy, sslAuthenticationOptions.EncryptionPolicy))
+            {
+                if (!CipherSuitesPolicyPal.WantsTls13(protocols))
+                {
+                    // We cannot provide neither TLS 1.3 or non TLS 1.3, user disabled all cipher suites
+                    throw new SslException(
+                        SR.Format(SR.net_ssl_encryptionpolicy_notsupported, sslAuthenticationOptions.EncryptionPolicy));
+                }
+
+                protocols = SslProtocols.Tls13;
+            }
+
+            if (sslAuthenticationOptions.CertificateContext != null && cacheSslContext)
+            {
+               sslAuthenticationOptions.CertificateContext.SslContexts.TryGetValue(sslAuthenticationOptions.EnabledSslProtocols, out sslCtx);
+            }
+
+            if (sslCtx == null)
+            {
+                // We did not get SslContext from cache
+                sslCtx = innerContext = AllocateSslContext(credential, sslAuthenticationOptions);
+            }
+
+            try
+            {
                 GCHandle alpnHandle = default;
                 try
                 {
+                    context = SafeSslHandle.Create(sslCtx, sslAuthenticationOptions.IsServer);
+                    Debug.Assert(context != null, "Expected non-null return value from SafeSslHandle.Create");
+                    if (context.IsInvalid)
+                    {
+                        context.Dispose();
+                        throw CreateSslException(SR.net_allocate_ssl_context_failed);
+                    }
+
+                    if (sslAuthenticationOptions.EncryptionPolicy != EncryptionPolicy.RequireEncryption)
+                    {
+                        // Sets policy and security level
+                        if (!Ssl.SetEncryptionPolicy(sslCtx, sslAuthenticationOptions.EncryptionPolicy))
+                        {
+                            throw new SslException( SR.Format(SR.net_ssl_encryptionpolicy_notsupported, sslAuthenticationOptions.EncryptionPolicy));
+                        }
+                    }
```

Comment:
Should we be doing this again if we already pulled the SSL_CTX out of the cache?

Since the cache didn't use this option as a key, are we potentially changing settings for some other connection?

Reply:
Looks like we won't ask the cache for an existing one if we need to hit this path.  Perhaps `Debug.Assert(!cacheSslContext);` inside the encryption policy check would help clarify that.